cl/citations/description_utility.py

```diff
@@ -0,0 +1,50 @@
+import re
+from math import log
+
+from cl.search.models import OpinionCluster
+
+_GERUND = r"(?:\S+ing)"
+_GERUND_THAT = rf"{_GERUND} that"
+_HOLDING = r"(?:holding|observing|deciding|ruling|recognizing)"
+_HOLDING_THAT = rf"{_HOLDING} that"
+
+OPTIMAL_NUM_WORDS = 20
+ALLOWED_NUM_WORDS_DEVIATION = 5
+
+
+def description_utility(
+    description: str, citing_opinion_cluster: OpinionCluster
+) -> int:
+    """
+    Takes a cleaned, non-spam description; returns an integer utility score.
+    """
+    # Baseline 500 so we don't end up with negative utility when we penalize
+    utility = 500
+    # "*ing that..." is the absolute gold standard
+    if re.match(_HOLDING_THAT, description):
+        utility += 400
+    if re.match(_GERUND_THAT, description):
+        utility += 300
+    # holding and synonyms are also pretty good
+    elif re.match(_HOLDING, description):
+        utility += 200
+    # in general gerunds are nice
+    elif re.match(_GERUND, description):
+        utility += 100
```

Comment:
Hm, is one of these better:

> Holding *that* the sky is blue

vs:

> Holding the sky is blue

The algo scores the first one much higher, but I'm not sure why. It makes sense to have words like "holding|observing|etc" rank higher than other gerunds, but I don't have the same instinct for the (in|ex)clusion of "that".

Reply:
I built this intuition after looking at some examples --- not because "that" always made descriptions better, but because "that" virtually guarantees that a helpful summary follows. Gerunds alone are also associated with summaries, but are not always followed by a nice summary. For example,

> holding that a State may not require a parade to include a group if the parade's organizer disagrees with the group's message

> remarking that examples of painting, music, and poetry are "unquestionably shielded"


as opposed to

>  relying upon common-law tradition

> referring to period of four weeks

> assuming, arguendo, that jurisdiction existed


Reviewing some of the data, however, it seems that a gerund (without 'that') parenthetical that is at least 10 words or so is equally likely to be as good as gerund + that; gerund + that is more likely to be helpful even if it is quite short. Perhaps we could reward gerund-only parentheticals equally to gerund + that when they are a bit longer.

Reply:
I'm going to have to lean on you guys for expertise here, TBH. So I think my comment can be taken or not. Whichever way we go, the more we can find ways to substantiate the assumptions that make it into the code, the better. I'm not sure how best to do that (maybe examples here, maybe examples in comments, something else?).

Reply:
Maybe more specific/directed test cases with a comment explaining the underlying rationale in the test case would help? And referring to these in the heuristic logic.